rcl_action/src/rcl_action/action_client.c

```diff
@@ -488,19 +521,47 @@ rcl_action_client_wait_set_get_entities_ready(
   RCL_CHECK_ARGUMENT_FOR_NULL(is_goal_response_ready, RCL_RET_INVALID_ARGUMENT);
   RCL_CHECK_ARGUMENT_FOR_NULL(is_cancel_response_ready, RCL_RET_INVALID_ARGUMENT);
   RCL_CHECK_ARGUMENT_FOR_NULL(is_result_response_ready, RCL_RET_INVALID_ARGUMENT);
-  if (2 != wait_set->size_of_subscriptions || 3 != wait_set->size_of_clients) {
-    RCL_SET_ERROR_MSG("wait set not initialized or not used by the action client alone");
-    return RCL_RET_WAIT_SET_INVALID;
-  }
-  *is_feedback_ready = !!wait_set->subscriptions[0];
-  *is_status_ready = !!wait_set->subscriptions[1];
-  *is_goal_response_ready = !!wait_set->clients[0];
-  *is_cancel_response_ready = !!wait_set->clients[1];
-  *is_result_response_ready = !!wait_set->clients[2];
+
+  const rcl_action_client_impl_t * impl = action_client->impl;
+  const size_t feedback_index = impl->wait_set_feedback_subscription_index;
+  const size_t status_index = impl->wait_set_status_subscription_index;
+  const size_t goal_index = impl->wait_set_goal_client_index;
+  const size_t cancel_index = impl->wait_set_cancel_client_index;
+  const size_t result_index = impl->wait_set_result_client_index;
+  if (feedback_index >= wait_set->size_of_subscriptions) {
+    RCL_SET_ERROR_MSG("wait set index for feedback subscription is out of bounds");
+    return RCL_RET_ERROR;
+  }
+  if (status_index >= wait_set->size_of_subscriptions) {
+    RCL_SET_ERROR_MSG("wait set index for status subscription is out of bounds");
+    return RCL_RET_ERROR;
+  }
+  if (goal_index >= wait_set->size_of_clients) {
+    RCL_SET_ERROR_MSG("wait set index for goal client is out of bounds");
+    return RCL_RET_ERROR;
+  }
+  if (cancel_index >= wait_set->size_of_clients) {
+    RCL_SET_ERROR_MSG("wait set index for cancel client is out of bounds");
+    return RCL_RET_ERROR;
+  }
+  if (result_index >= wait_set->size_of_clients) {
+    RCL_SET_ERROR_MSG("wait set index for result client is out of bounds");
+    return RCL_RET_ERROR;
+  }
+
+  const rcl_subscription_t * feedback_subscription = wait_set->subscriptions[feedback_index];
+  const rcl_subscription_t * status_subscription = wait_set->subscriptions[status_index];
+  const rcl_client_t * goal_client = wait_set->clients[goal_index];
+  const rcl_client_t * cancel_client = wait_set->clients[cancel_index];
+  const rcl_client_t * result_client = wait_set->clients[result_index];
+  *is_feedback_ready = (&impl->feedback_subscription == feedback_subscription);
+  *is_status_ready = (&impl->status_subscription == status_subscription);
+  *is_goal_response_ready = (&impl->goal_client == goal_client);
+  *is_cancel_response_ready = (&impl->cancel_client == cancel_client);
+  *is_result_response_ready = (&impl->result_client == result_client);
```

Comment:
@jacobperron though extremely unlikely, there's the possibility that e.g. `feedback_subscription` is neither `NULL` nor `&impl->feedback_subscription`. Just pointing it out in case we may want to return in error instead.